### PR TITLE
Update SchemaFactory to allow the annotation cache dir to be configurable

### PR DESCRIPTION
**Pull request by MattBred**
Addresses https://github.com/thecodingmachine/graphqlite/issues/309

### CODE REVIEW

**Comment by moufmouf** [comment]
Hey @MattBred ,

Thanks a lot for this PR. If you don't mind, I'll keep it open for now.
I'm planning on working to add support for PHP 8 attributes, and I'll need some cache for those. The cache might be very different from the Doctrine cache used for Doctrine annotations, and I'm wondering if I should not mutualize both caches (meaning I would completely remove the Doctrine cache for annotations and replace it with a PSR-16 cache used elsewhere in the project...)

Not sure yet, but until I make my mind, I would like to avoid adding an additional parameter to the SchemaFactory constructor (since that parameter might disappear soon!)

**Comment by MattBred** [comment]
Sounds good @moufmouf .

I've found a few places in graphqlite + the symfony bundle that use the system temp dir instead of a configurable dir, so this PR was to address that (among some others I made)

**Comment by mdoelker** [comment]
Is there a good reason this does not use the cache instance provided via the constructor? This caught me by surprise quite a bit.

**Comment by moufmouf** [comment]
@mdoelker yes, there is (unfortunately!)

The cache for the annotation is a Doctrine Cache (that is not PSR-16 compatible, nor PSR-6 compatible).
It is also very fast. And we need a cache that is very fast, because we typically want very high performance for annotation fetching (we can hit this cache a big number of times per requests).

We could of course use an adapter to turn the Doctrine cache into a PSR-16 cache. The fact is that the PSR-16 cache is typically slower (if you look at the Symfony implementation, it needs to check if the key is valid according to PSR-16 rules, etc...). So it is solid, but I fear we would loose a few milliseconds here.

Moreover, I don't want users to configure a remote cache (like Redis or a DB backed cache) for annotations. That would completely destroy the performances of GraphQLite. So the only 2 viable options are APCu and PHPFile cache. So I decided to hard code those in the SchemaFactory.

File: src/SchemaFactory.php [hunk 1]
```diff
@@ -109,11 +111,12 @@ class SchemaFactory
     /** @var string */
     private $cacheNamespace;
 
-    public function __construct(CacheInterface $cache, ContainerInterface $container)
+    public function __construct(CacheInterface $cache, ContainerInterface $container, ?string $annotationCacheDir = null)
```
Comment (moufmouf) [inline src/SchemaFactory.php:114]: @MattBred Would you mind moving the `$annotationCacheDir ` parameter into a setter?

The `SchemaFactory` class is typically following the "builder" pattern and any non compulsory parameter is configured via a setter.

**Review by moufmouf — CHANGES_REQUESTED** [review]
Hey @MattBred ,

I'm :+1: on merging this once you move the `$annotationCacheDir` parameter into a setter (see comment above)

Thanks a lot for your work!

File: src/SchemaFactory.php (hunk 1 above)
Comment (MattBred) [reply src/SchemaFactory.php:114]: Sounds good, I will take care of that today

Comment (MattBred) [reply src/SchemaFactory.php]: @moufmouf Updated

**Comment by mdoelker** [comment]
Thanks for the explanation and for proceeding with this. Btw, does it make sense to check the `apcu_enabled()` func? Not too familiar with APCu.

```php
$cache = function_exists('apcu_enabled') && apcu_enabled() ? new ApcuCache() : new PhpFileCache($cacheDir . '/graphqlite.' . crc32(__DIR__));
```

`Returns TRUE when APCu is usable in the current environment, FALSE otherwise.`

**Comment by moufmouf** [comment]
@mdoelker It does make a lot of sense, indeed!

Do you want to make add a comment with this suggestion in this PR so that @MattBred can merge it?

File: src/SchemaFactory.php [hunk 2]
```diff
@@ -213,7 +222,8 @@ private function getDoctrineAnnotationReader(): Reader
             AnnotationRegistry::registerLoader('class_exists');
             $doctrineAnnotationReader = new DoctrineAnnotationReader();
 
-            $cache = function_exists('apcu_fetch') ? new ApcuCache() : new PhpFileCache(sys_get_temp_dir() . '/graphqlite.' . crc32(__DIR__));
+            $cacheDir = $this->annotationCacheDir ?? sys_get_temp_dir();
+            $cache = function_exists('apcu_fetch') ? new ApcuCache() : new PhpFileCache($cacheDir . '/graphqlite.' . crc32(__DIR__));
```
Comment (mdoelker) [inline src/SchemaFactory.php:226]: We can improve the APCu check here

```suggestion
            $cache = function_exists('apcu_enabled') && apcu_enabled() ? new ApcuCache() : new PhpFileCache($cacheDir . '/graphqlite.' . crc32(__DIR__));
```

**Comment by MattBred** [comment]
@moufmouf Have a look now, should be good to go?

**Comment by moufmouf** [comment]
Excellent! Thanks a lot for your contribution!

**Pull request merged**